src/cascadia/TerminalSettingsModel/ActionMap.h

```diff
@@ -0,0 +1,118 @@
+/*++
+Copyright (c) Microsoft Corporation
+Licensed under the MIT license.
+
+Module Name:
+- ActionMap.h
+
+Abstract:
+- A mapping of key chords to actions. Includes (de)serialization logic.
+
+Author(s):
+- Carlos Zamora - September 2020
+
+--*/
+
+#pragma once
+
+#include "ActionMap.g.h"
+#include "IInheritable.h"
+#include "Command.h"
+#include "../inc/cppwinrt_utils.h"
+
+// fwdecl unittest classes
+namespace SettingsModelLocalTests
+{
+    class KeyBindingsTests;
+    class DeserializationTests;
+    class TerminalSettingsTests;
+}
+
+namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+{
+    typedef size_t InternalActionID;
+
+    struct KeyChordHash
+    {
+        std::size_t operator()(const Control::KeyChord& key) const
+        {
+            return ::Microsoft::Terminal::Settings::Model::HashUtils::HashProperty(key.Modifiers(), key.Vkey());
+        }
+    };
+
+    struct KeyChordEquality

```

Comment:
Same as above, this was stolen from `KeyMapping` and is used in the same declarations for key map.

This one makes more sense to just have as a global operator though, so I'll do that.

Reply:
Ugh, I tried just that, but then all the key bindings were broken. It seems like these `unordered_map`s _need_ `KeyChordEquality`.

Reply:
the unordered map should be using std::equals or std::is_equal which will use `operator==` from the type _unless it's overridden in the template args_.

Like, this should work: (writing code in godbolt)

Reply:
okay, this works. it actually requires the operator to be in the leaf namespace. i could have sworn it should work in the parent namespace.

```c++
#include <unordered_map>

namespace A::B::C {
    struct Thing {
        int a, b;
    };
    // it apparently has to go here
    inline bool operator==(const C::Thing& lhs, const C::Thing& rhs) {
        return lhs.a==rhs.a && lhs.b==rhs.b;
    }
}

namespace A::B {
    struct ThingHasher {
        size_t operator()(const C::Thing& lhs) const { return 7; }
    };
    // okay, i thought you could put it here and because A::B::C is inside A::B it would find it.
}

int fleh() {
    std::unordered_map<A::B::C::Thing, int, A::B::ThingHasher> augh;
    augh[A::B::C::Thing{1,2}] = 3;
    augh[A::B::C::Thing{3,4}] = 7;
    return augh.find(A::B::C::Thing{1,2}) == augh.end();
}
```

Reply:
Ah, ADL only uses the leafmost namespace for a lot of things. So yes: if you introduce an `operator==` into the MS.Terminal.Control namespace in your header file _it will get used automatically_

Reply:
broken = not recognized. Like, "ctrl+shift+space" wouldn't be detected.

Reply:
llllet's just leave it. we can fix it in post.